arangod/Cluster/AgencyCache.cpp

```diff
@@ -0,0 +1,300 @@
+  : Thread(server, "AgencyCache"), _commitIndex(0),
+    _readDB(server, nullptr, "raadDB"), _callbackRegistry(callbackRegistry) {}
+
+
+AgencyCache::~AgencyCache() {}
+
+/// Start all agent thread
+bool AgencyCache::start() {
+  LOG_TOPIC("9a90f", DEBUG, Logger::AGENCY) << "Starting agency cache worker.";
+  Thread::start();
+  return true;
+}
+
+// Get Builder from readDB mainly /Plan /Current
+std::tuple <query_t, index_t> const AgencyCache::get(
+  std::string const& path) const {
+  std::lock_guard g(_storeLock);
+  auto ret = std::make_shared<VPackBuilder>();
+  if (_commitIndex > 0) {
+    _readDB.get("arango/" + path).toBuilder(*ret);
+  }
+  return std::tuple(ret, _commitIndex);
+}
+
+// Get Builder from readDB mainly /Plan /Current
+std::tuple <query_t, index_t> const AgencyCache::get(
+  std::vector<std::string> const& paths) const {
+  std::lock_guard g(_storeLock);
+
+  auto result = std::make_shared<arangodb::velocypack::Builder>();
+  if (_commitIndex > 0) {
+    auto query = std::make_shared<arangodb::velocypack::Builder>();
+    {
+      VPackArrayBuilder outer(query.get());
+      VPackArrayBuilder inner(query.get());
+      for (auto const& i : paths) {
+        query->add(VPackValue(i));
+      }
+    }
+    _readDB.read(query, result);
+  }
+  return std::tuple(result, _commitIndex);
+}
+
+futures::Future<arangodb::Result> AgencyCache::waitFor(index_t index) {
+  std::lock_guard s(_storeLock);
+  if (index <= _commitIndex) {
+    return futures::makeFuture(arangodb::Result());
+  }
+  std::lock_guard w(_waitLock);
+  return _waiting.emplace(index, futures::Promise<arangodb::Result>())->second.getFuture();
+}
+
+index_t AgencyCache::index() const {
+  std::lock_guard g(_storeLock);
+  return _commitIndex;
+}
+
+void AgencyCache::run() {
+
+  using namespace std::chrono;
+  TRI_ASSERT(AsyncAgencyCommManager::INSTANCE != nullptr);
+
+  _commitIndex = 0;
+  _readDB.clear();
+  double wait = 0.0;
+  VPackBuilder rb;
+  std::vector<uint32_t> toCall;
+  // Long poll to agency
+  auto sendTransaction =
+    [&](VPackBuilder& r) {
+      index_t commitIndex = 0;
+      {
+        std::lock_guard g(_storeLock);
+        commitIndex = _commitIndex;
+      }
+      auto ret = AsyncAgencyComm().poll(60s, commitIndex+1);
+      //r.add(ret.value());
+      return ret;
+    };
+
+  // while not stopping
+  //   poll agency
+  //   if result
+  //     if firstIndex == 0
+  //       overwrite readDB, update commitIndex
+  //     else
+  //       apply logs to local cache
+  //   else
+  //     wait ever longer until success
+
+  while (!this->isStopping()) {
+
+    rb.clear();
+    toCall.clear();
+    std::this_thread::sleep_for(std::chrono::duration<double>(wait));
+    
+    auto ret = sendTransaction(rb)
+      .thenValue(
+        [&](AsyncAgencyCommResult&& rb) {
+          if (rb.ok() && rb.statusCode() == arangodb::fuerte::StatusOK) {
+            auto tmp = VPackParser::fromJson(rb.payloadAsString());
```

Comment:
Why? If the connections is VST you will get a valid vpack as response. You can just call `rb->slice()`, right?

Reply:
fixed.